robsize.cc

```diff
@@ -416,56 +421,85 @@ void print_tests() {
-        } else if (!strcmp(argv[i], "--list")) {
-            print_tests();
-            exit(EXIT_SUCCESS);
-        } else if (!strcmp(argv[i], "--csv")) {
-            plot_mode = true;
-        } else if (!strcmp(argv[i], "--slow")) {
-            outer_its <<=1;
-        } else if (!strcmp(argv[i], "--superfast")) {
-            its >>=4;
-            outer_its >>=3;
-        } else if (!strcmp(argv[i], "--write-asm")) {
-            // print the generated instructions to a file and quit
-            print_ibuf = true;
-        } else {
-            fprintf(stderr, "Uncognized argument: %s\n", argv[i]);
-            return false;
+/* Command line options for getopt_long() */
+static struct option long_options[] = {
+    {"help",       no_argument, NULL, 'h'},
+    {"list",       no_argument, NULL, 'h'},
+    {"csv",        no_argument, NULL, 'c'},
+    {"write-asm",  no_argument, NULL, 'w'},
+    {"slow",       no_argument, NULL, 's'},
+    {"fast",       no_argument, NULL, 'f'},
+    {"superfast",  no_argument, NULL, 'g'},
+    {"start", required_argument, NULL, 'i'},
+    {"stop",  required_argument, NULL, 'j'},
+    {0, 0, 0, 0}
+};
+
+void handle_args(int argc, char *argv[]) {
+    int optval = 0;
+    int opt_idx = 0;
+    while ((optval = getopt_long(argc, argv, "", long_options, &opt_idx)) >= 0) {
+        switch (optval) {
+            case 'h': /* help */
+                print_usage();
+                exit(EXIT_SUCCESS);
+                break;
+            case 'c': /* csv */
+                plot_mode = true;
+                break;
+            case 'w': /* write-asm */
+                print_ibuf = true;
+                break;
+            case 's': /* slow */
+                outer_its <<= 1;
+                break;
+            case 'f': /* fast */
+                its >>= 2;
+                outer_its >>= 2;
+                break;
+            case 'g': /* super-fast */
+                its >>= 4;
+                outer_its >>= 3;
+                break;
+            case 'i': /* start */
+                if (sscanf(optarg, "%d", &start_icount) <= 0) {
+                    fprintf(stderr, "Unrecognized value for --start: %s\n",  optarg);
+                    print_usage();
+                    exit(EXIT_FAILURE);
+                }
+                break;
+            case 'j': /* stop */
+                if (sscanf(optarg, "%d", &stop_icount) <= 0) {
+                    fprintf(stderr, "Unrecognized value for --stop: %s\n",  optarg);
+                    print_usage();
+                    exit(EXIT_FAILURE);
+                }
+                break;
+            default:
+                print_usage();
+                exit(EXIT_FAILURE);
         }
     }
 
-    return true;
-}
+    // At most one non-option argument is the positional arg for the test ID
+    // getopt moves this to the end, so we cannot sue argv[1] here!
+    if (optind < argc) {
+        optind += (sscanf(argv[optind], "%d", &instr_type) > 0);
```

Comment:
If the scanf fails, we don't detect it?

Reply:
No because if it failed then we did not have a positional argument for the test ID and therefore we can use the default test ID of 4 instead.

Reply:
Ah right, I missed that part where if it fails `optind` won't be incremented we will fail in the next check.

Reply:
I figured as long as the rest of the code uses C idioms I might as well use the idiomatic trickery that comes with the language :p